### PR TITLE
test: add mutation testing across workspaces

### PR DESCRIPTION
**Pull request by hiroppy**
# Summary

- Add Stryker mutation testing to the six workspaces that own unit tests, with explicit targets, exclusions, thresholds, Vitest integration, and TypeScript checking.
- Run full mutation suites as independent `fail-fast: false` matrix jobs and upload verified HTML/JSON reports.
- Enable Stryker incremental mode and cache each workspace report by OS, workspace, dependency/config fingerprint, and commit SHA. Invalid cached JSON falls back to a full run.
- Preserve TypeScript 7 application checks while using a scoped TypeScript 6 compiler API alias for Stryker 9.6.1.

## Linear Issue

- [HIR-113](https://linear.app/hiroppy/issue/HIR-113/mutation-testing%E3%82%92%E5%AE%9F%E6%96%BD%E3%81%99%E3%82%8B)

### Selected Quality Characteristics

| Quality characteristic | Why it is relevant | Acceptance criteria |
| ---------------------- | ------------------ | ------------------- |
| Functional suitability | Reused results must retain the same mutation-quality decision as a cold run | Cold and warm runs produce the same scores and meet all six thresholds |
| Performance efficiency | Full mutation feedback previously required tens of minutes per run | An unchanged warm run reuses results and materially reduces elapsed time |
| Reliability | Missing, stale, or corrupt cache data must not bypass the gate | Cache misses and invalid JSON fall back to full evaluation; reports remain required and non-empty |
| Maintainability | Cache invalidation and workspace scope must be reviewable | Workspace configs and dependency/config fingerprint inputs remain explicit |
| Security | CI needs read-only repository access | Workflow grants only `contents: read` and checkout does not persist credentials |

### Test Techniques

| Test technique | Selection rationale | Expected coverage |
| -------------- | ------------------- | ----------------- |
| State transition testing | Incremental behavior depends on prior report and file state | Cold, warm unchanged, test-changed, cache-miss, and invalid-cache states |
| Decision table testing | Reuse depends on workspace, fingerprint, and commit key combinations | Per-workspace cache isolation and dependency/config invalidation |
| Boundary value analysis | Date logic has timezone and calendar boundaries | Existing mutation-focused JST, month/year rollover, and invalid-input tests |
| Branch testing | Surviving mutants expose indistinguishable branches | Stryker evaluates branch/operator changes against each unit suite |
| Error guessing | Cache corruption and missing report artifacts are operational failure risks | JSON validation, full fallback, and non-empty artifact checks |

### Primary Test Conditions

- A cold run without incremental files executes all six mutation suites and creates incremental, HTML, and JSON reports.
- An unchanged warm run reuses prior mutant results while still running Stryker's initial test run and threshold/report checks.
- A changed Vitest test file invalidates the affected file's results; invalid cached JSON is removed before Stryker and triggers a full run.
- Dependency, patch, package, Stryker, Vitest, or TypeScript config changes alter the cache fingerprint.
- Pull request, scheduled, and manual workflows retain independent matrix execution, `fail-fast: false`, 90-minute timeouts, and report uploads.

### Out-of-Scope Risks

- Stryker documents that changes outside mutated/test files and environment changes are not detected internally; dependency and relevant config files are therefore included in the CI fingerprint. Runtime secrets are not used by these unit mutation suites.
- Browser-context scrapers, Monte Carlo hot paths, schema/seed/type-only files, simulator, and real-database integration remain outside the deterministic unit mutation gate.

## Verification

- [x] Unit tests
- [x] Type checking
- [x] Lint
- [x] Storybook tests
- [ ] E2E tests
- [x] Manual verification

### Commands Executed

```text
pnpm test — passed, 6/6 workspace tasks (web 1015, crawler 368, db 400 + 32 skipped)
pnpm turbo typecheck — passed, 8/8 workspaces
pnpm lint — passed
pnpm format:check — passed
pnpm knip — passed
six workspace Stryker dry runs — passed
pnpm test:mutation (cold) — passed, 6/6 in 23m50s; all thresholds and 18 report files passed
pnpm test:mutation (warm unchanged) — passed, 6/6 in 45.38s; same scores and reports, approximately 96.8% faster
meta cold/warm/test-changed/invalid-cache state proofs — passed
workflow YAML parse and git diff --check — passed
```

### Checks Not Run

- E2E tests — the change only affects mutation-test configuration and CI caching; existing unit/Storybook coverage and the PR CI E2E jobs provide regression coverage.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -21,6 +21,17 @@ jobs:
       - run: pnpm knip
       - run: pnpm typecheck --affected
       - run: pnpm test --affected
+      - run: pnpm test:mutation
```
Comment (hiroppy) [inline .github/workflows/ci.yml]: `pnpm test:mutation` now runs the full monorepo mutation suite in the primary PR test job. In a fresh checkout I verified `pnpm --filter @mf-dashboard/date-utils test:mutation` passes, but the root `pnpm test:mutation` had only completed `date-utils` and `meta` after 10 minutes and was still inside the `db` workspace with 2,607 mutants and multi-hour remaining estimates; a targeted crawler run also exceeded 6 minutes before completion. This will make every PR's main CI feedback path impractically slow/flaky. Please move full mutation testing to a separate scheduled/manual job or constrain PR CI to affected/targeted workspaces with explicit concurrency/timeouts before gating all PRs on it.

Comment (hiroppy) [reply .github/workflows/ci.yml]: Addressed in b190c8a. The full suite now runs in a separate daily/manual Mutation Testing workflow with a 90-minute job timeout, while the primary PR CI retains the affected typecheck and test feedback path. Mutation reports are still uploaded for inspection. Verified with pnpm format:check and git diff --check; the repository does not provide actionlint, so the pushed workflow run will provide GitHub syntax validation.

File: apps/web/package.json [hunk 2]
```diff
@@ -9,6 +9,7 @@
     "build:demo": "DB_PATH=../../data/demo.db DEMO_MODE=true next build",
     "start": "next start",
     "test": "vitest run",
+    "test:mutation": "stryker run",
     "test:unit": "vitest run --project unit",
```
Comment (hiroppy) [inline apps/web/package.json:13]: `@mf-dashboard/web`'s new mutation script runs `stryker` without first generating Next's global `PageProps`/`LayoutProps` types. In the disposable checkout, `CI=true pnpm --filter @mf-dashboard/web test:mutation` fails during Stryker's TypeScript checker with TS2304 for `PageProps`/`LayoutProps` before running mutants, so the scheduled root `pnpm test:mutation` workflow will fail as soon as it reaches the web workspace. Please generate Next types (for example `next typegen`) before Stryker or adjust the mutation TypeScript checker inputs.

Comment (hiroppy) [reply apps/web/package.json:13]: Addressed in e3b8bc8. The web test:mutation script now runs next typegen before Stryker, so a fresh scheduled checkout creates PageProps/LayoutProps before the TypeScript checker starts. Verified next typegen plus stryker run --dryRunOnly: the checker initialized and all 480 dry-run tests passed; pnpm format:check and git diff --check also passed.

File: turbo.json [hunk 3]
```diff
@@ -22,6 +22,9 @@
     "test": {
       "dependsOn": ["^build"]
     },
+    "test:mutation": {
```
Comment (hiroppy) [inline turbo.json:25]: `test:mutation` is wired through `dependsOn: ["^test:mutation"]`, so the scheduled `pnpm test:mutation` runs mutation suites as a dependency chain rather than as independent report producers. If a foundational workspace such as `date-utils`, `db`, or `analytics` drops below its threshold, Turbo fails before downstream suites like `web`/`crawler` can run and upload their reports, so the daily workflow no longer gives full-monorepo mutation visibility. Prefer a workspace matrix with `fail-fast: false` (or otherwise run each package mutation script independently) and aggregate the results/artifacts at the workflow level.

File: apps/web/stryker.config.json [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  "plugins": ["@stryker-mutator/typescript-checker", "@stryker-mutator/vitest-runner"],
+  "reporters": ["clear-text", "progress", "html", "json"],
+  "testRunner": "vitest",
+  "thresholds": {
+    "high": 80,
+    "low": 55,
+    "break": 55
+  },
+  "vitest": {
+    "configFile": "vitest.mutation.config.ts"
```
Comment (hiroppy) [inline apps/web/stryker.config.json:43]: `@mf-dashboard/web` の `test:mutation` がまだ Stryker 実行中に落ちる。`CI=true pnpm --filter @mf-dashboard/web test:mutation` は `next typegen` と initial dry run までは通るが、その後 Vitest が sandbox 内で `vitest.mutation.config.ts` を解決できず `[UNRESOLVED_ENTRY] Cannot resolve entry module vitest.mutation.config.ts` で失敗する。Stryker がコピーする対象にこの config を含める（例: `files` に追加）か、sandbox 内で解決できる既存 config を参照して、scheduled `pnpm test:mutation` が web workspace で完走するようにしてほしい。

File: turbo.json (hunk 3 above)
Comment (hiroppy) [reply turbo.json:25]: Addressed in f674124. The scheduled/manual workflow now uses a six-workspace matrix with `fail-fast: false`, runs each package mutation script directly, and uploads a uniquely named report artifact from that workspace even when another matrix job fails. The Turbo mutation task no longer chains through upstream workspaces and is non-cacheable. Validated all six package mappings, workflow YAML, the full web mutation suite, and the regular test/typecheck/lint/format/knip gates.

File: apps/web/stryker.config.json (hunk 4 above)
Comment (hiroppy) [reply apps/web/stryker.config.json:43]: Addressed in f674124. The web Stryker sandbox now explicitly re-includes `.next/types`, `src`, `tsconfig.json`, and `vitest.mutation.config.ts` through the current `ignorePatterns` configuration. Verified with an actual four-mutant run (100% score), then the full web suite: 2,465 mutants, 485 initial tests, 61.56% score against the 55% break threshold, completed in 9m30s without config resolution errors.

File: .github/workflows/mutation.yml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+on:
+  pull_request:
+  schedule:
+    - cron: "0 15 * * *"
+  workflow_dispatch:
+
+concurrency:
+  group: mutation-testing-${{ github.event_name == 'pull_request' && github.event.pull_request.number || 'full' }}
+  cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply .github/workflows/mutation.yml:14]: Addressed in eb4cc6f. The mutation workflow now declares workflow-level `contents: read`; all unspecified token scopes are disabled. The existing pull request, schedule, and manual triggers and dynamic concurrency group remain unchanged. The same change splits the full suite into a six-workspace `fail-fast: false` matrix after the previous single job reached its 90-minute timeout. Validated the workflow YAML and matrix mappings, all six Stryker dry runs, a full meta mutation run, and the regular test/typecheck/lint/format/knip gates.

File: .github/workflows/mutation.yml [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+          - workspace: date-utils
+            path: packages/date-utils
+          - workspace: db
+            path: packages/db
+          - workspace: meta
+            path: packages/meta
+    runs-on: ubuntu-latest
+    timeout-minutes: 90
+    steps:
+      - uses: actions/checkout@v7.0.1
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply .github/workflows/mutation.yml:37]: Addressed in c897f44: checkout now sets `persist-credentials: false`, and no later step requires Git authentication.

File: .github/workflows/mutation.yml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      - uses: ./.github/actions/setup-node
+      - run: pnpm --filter @mf-dashboard/${{ matrix.workspace }} test:mutation
+      - if: ${{ !cancelled() }}
+        uses: actions/upload-artifact@v7.0.1
+        with:
+          name: mutation-report-${{ matrix.workspace }}
+          path: |
+            ${{ matrix.path }}/reports/mutation/mutation.html
+            ${{ matrix.path }}/reports/mutation/mutation.json
+          retention-days: 7
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply .github/workflows/mutation.yml:69]: Addressed in c897f44: the workflow now verifies both report files with `test -s` before upload and sets `if-no-files-found: error`.

File: apps/web/stryker.config.json [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  "checkers": ["typescript"],
+  "coverageAnalysis": "perTest",
+  "ignorePatterns": [
+    "**",
+    "!.next/types/**/*.ts",
+    "!next-env.d.ts",
+    "!package.json",
+    "!src/**/*",
+    "!tsconfig.json",
+    "!vitest.mutation.config.ts"
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/web/stryker.config.json:14]: This does not reproduce. The completed web mutation CI job on c44bb5a reported `Initial test run succeeded. Ran 508 tests`, and after merging the latest main the local `stryker run --dryRunOnly` again completed all 508 tests successfully. The selected tests and their support files are therefore available in Stryker’s sandbox; changing the whitelist would add unnecessary inputs.
